### DISCUSSION TITLE
[Security] enable secrets to be read from a file and not directly from an environment variable

### DISCUSSION THREAD

**Discussion by happysalada** [meilisearch/product#201] [Feedback & Feature Proposal]
As far as I know, there is one secret to supply to meilisearch and it's the `MEILI_MASTER_KEY`. That environment variable is read directly. To fit security best practices, secrets should be read from a file. (More details below)

Whether using [docker secrets](https://docs.docker.com/engine/swarm/secrets/) or [systemd loadcredentials](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#LoadCredential=ID:PATH), tools managing service pass secrets via files (not an environment variables). [some reasons are detailed here](https://diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/)

I'm happy to go into more details onto why this might not be a good idea.

I understand that this might not make on the top of the priority list, but just something to keep in mind for when you have time.

let me know if this needs more information

**Comment by curquiza** [comment]
Hello @happysalada! Thanks for your feature request, I move this to the right repo (cf this [part in the README](https://github.com/meilisearch/MeiliSearch#feature-request)) 🙂
This might be a valuable addition!

**Reply by happysalada** [reply]
Thank you for your help!

**Comment by gmourier** [comment]
Hello @happysalada 👋

Thanks for the feedback and the suggestion. I would love to hear more details about this! 

It is indeed more secure and avoids dumping the whole env vars in another process. Would having a configuration file to boot the instance also make sense? I try to see if this subject can be treated with a need that we already have which is to have a configuration file, I'm also thinking about the beginner who could have some difficulties with systemd or docker secrets.

Thanks

**Reply by happysalada** [reply]
a docker secret takes an individual file, and systemd's load credential works essentially the same.

You will need to change the way the variable is read. Rather than directly taking the value from an environment variable. The environment variable for secrets should contain a path, and then the code should read the file from the path.

In order to provide a smooth transition, You could add one more environment variable (As far as I know, the only secret is the master key). So the change would be to add an environment variable MASTER_KEY_PATH and inside the code do something like
```
if MASTER_KEY_PATH {
    master_key = read file MASTER_KEY_PATH
} else if MASTER_KEY {
   master_key = MASTER_KEY
} else {
   panic "master key needed"
}
```

This is peudo code, but I don't believe the change would be bigger than that.
I could be wrong if there are more secrets that need to be passed.

You could batch this with adding a config file, but ultimately the change is not very related. ( I completely understand that this change needs to be justified business wise though)

**Comment by failable** [comment]
Is docker file secret supported now?

**Reply by curquiza** [reply]
Hello @liebkne 
No, we have not implemented anything so far for this on the engine side, and it's not planned currently

However we have implemented the possibility to add use [a config file](https://www.meilisearch.com/docs/learn/configuration/instance_options#configuration-file)
